### PR TITLE
Add `bundle_url` support for remote OPA policy bundles

### LINKED ISSUE #228: Feature Request: Load policies from a remote OPA bundle server
### Summary

Add support for loading Rego policies from a remote HTTP bundle server at plugin startup, as an alternative to the existing `policy_dir` local filesystem option.

### Motivation

In team and CI environments, policies are typically centrally managed and distributed via an OPA bundle server (S3, GCS, nginx, or a custom registry). Today, every consumer of `tflint-ruleset-opa` must either:

- Commit policy files into each repository and keep them in sync, or
- Pre-download the bundle as a separate CI step before running `tflint`

Neither approach is ideal. A native `bundle_url` option would let the plugin fetch the authoritative policy bundle directly at startup, making centralized policy distribution a first-class concern.

### Proposed Configuration

```hcl
plugin "opa" {
  enabled  = true
  version  = "0.10.0"
  source   = "github.com/terraform-linters/tflint-ruleset-opa"

  bundle_url   = "https://policy-server.example.com/bundles/tflint.tar.gz"
  bundle_token = "..."  # optional, bearer token for auth
}
```

The `bundle_token` value should also be resolvable from an environment variable to avoid hardcoding secrets.

### Proposed Behaviour

- If `bundle_url` is set, the plugin fetches the bundle over HTTP(S) at startup, parses it with `bundle.NewReader()`, and passes it to `rego.New()` via `rego.ParsedBundle()` — exactly the same in-process evaluation path used today
- If `bundle_url` is not set, behaviour is identical to today (`policy_dir` is used)
- Both can coexist: `bundle_url` supplies shared org-wide policies, `policy_dir` supplies local overrides
- Optional: respect `ETag` / `If-None-Match` for caching, to avoid re-downloading an unchanged bundle on every `tflint` run

### What This Is Not

This is **not** a request for the full OPA bundle polling runtime (`plugins/bundle` + plugin manager). That machinery is designed for long-lived server processes. For a CLI tool like `tflint`, a single one-shot HTTP fetch at startup is the right model — simple, fast, and with no persistent state.

### Implementation Notes

The change is self-contained and relatively small:

- Extend `Config` with `BundleURL` and `BundleToken` fields
- In `ApplyConfig`, branch on whether `BundleURL` is set: if so, `http.Get` the URL, pipe the response body through `bundle.NewReader().Read()`, and pass the result to `NewEngine` alongside or instead of the `loader.Result`
- `NewEngine` (or a sibling constructor) accepts a `*bundle.Bundle` and passes it to `rego.New()` via `rego.ParsedBundle()`
- The custom Rego builtins (`terraform.resources`, `tflint.issue`, etc.) are registered on the `rego.Rego` instance as today — they work regardless of where the policies came from

Roughly ~100 lines of new Go, no new dependencies (all required packages are already present in the OPA module).

### Alternatives Considered

- **Pre-downloading the bundle in CI**: works but requires an extra step and couples the CI script to the bundle server URL/auth, rather than keeping that in `.tflint.hcl`
- **Committing policies to each repo**: creates drift and makes org-wide policy updates require PRs across every repository

### Are you willing to contribute a PR?

Yes, if the approach is agreed upon.

### PR DESCRIPTION
**Pull request by mostealth**
## Summary

Add support for loading Rego policies from a remote HTTP bundle server at plugin startup, as an alternative to the existing `policy_dir` local filesystem option.

In team and CI environments, policies are typically centrally managed and distributed via an OPA bundle server (S3, GCS, nginx, or a custom registry). This feature lets the plugin fetch the authoritative policy bundle directly at startup, making centralized policy distribution a first-class concern.

## Changes

- Add `bundle_url` config attribute and `TFLINT_OPA_BUNDLE_URL` environment variable
- Fetch bundle via HTTP(S) and parse with OPA's `bundle.NewReader`
- Cache bundles locally at `~/.tflint.d/cache/opa/bundles` with ETag support to avoid redundant downloads
- Treat `bundle_url` and `policy_dir` as mutually exclusive: if a bundle is configured and the local policy directory contains any policies or data, the plugin errors out (only an empty or absent local directory is ignored in favor of the bundle)
- Refactor `NewEngine` to accept `storage.Store` and `map[string]*ast.Module` directly
- Size-limit bundle downloads to 256 MB with a clear error message
- Add unit tests for `fetchBundle`, `bundleURL` config, and ETag caching (including cache corruption fallback), and the `bundle_url`/`policy_dir` conflict path
- Add an integration test for the remote bundle scenario
- Document `bundle_url`, `TFLINT_OPA_BUNDLE_URL`, and the `bundle_url`/`policy_dir` mutual exclusion

## Configuration

```hcl
plugin "opa" {
  enabled    = true
  bundle_url = "https://policy-server.example.com/bundles/tflint.tar.gz"
}
```

`bundle_url` and `policy_dir` are mutually exclusive — configuring a bundle alongside a local policy directory that contains policies or data is an error.

## Test plan

- Unit tests: go test ./opa -v (fetchBundle, bundleURL config, ETag caching, cache fallback, bundle_url/policy_dir conflict)
- Integration tests: go test ./integration -v (remote bundle)
- Existing tests pass unchanged

## Related

Closes #228

### CODE REVIEW

File: opa/bundle.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+var maxBundleSize int64 = 256 << 20 // 256 MB
+
+type limitedReader struct {
+	r         io.Reader
+	remaining int64
+}
+
+func (l *limitedReader) Read(p []byte) (int, error) {
+	if l.remaining <= 0 {
+		return 0, fmt.Errorf("bundle exceeds maximum size (%d MB)", maxBundleSize>>20)
```
Comment (wata727) [inline opa/bundle.go]: `io.ReadAll` performs one extra read after consuming the final chunk. If the response body is exactly `maxBundleSize` bytes long, remaining reaches 0 and the next call returns bundle exceeds maximum size instead of `io.EOF`, so bundles at the documented 256 MiB limit are rejected.

This only shows up for archives that land exactly on the threshold, but the current check makes the effective limit one byte smaller than intended.

File: opa/ruleset.go [hunk 2]
```diff
@@ -40,23 +44,79 @@ func (r *RuleSet) ApplyConfig(body *hclext.BodyContent) error {
...
+		for k, m := range ret.ParsedModules() {
+			localModules[k] = m
+		}
+		s, err := ret.Store()
+		if err != nil {
+			return fmt.Errorf("failed to create policy store; %w", err)
+		}
+		// When local data files exist, they take precedence over bundle data.
+		if len(ret.Documents) > 0 {
+			store = s
```
Comment (wata727) [inline opa/ruleset.go]: When `bundle_url` is used together with a local `policy_dir` that contains any JSON/YAML documents, this branch discards the entire bundle `data` tree and keeps only `ret.Store()`. That breaks bundle policies which rely on bundle-provided data but are otherwise meant to coexist with local overrides.

For example, a bundle shipping shared lookup tables under `data.shared` will stop working as soon as a team adds an unrelated local data file, even though the comment says local data should only take precedence.

File: docs/configuration.md [hunk 3]
```diff
@@ -24,3 +25,29 @@ Change the directory from which policies are loaded. The priority is as follows:
...
+```hcl
+plugin "opa" {
+  enabled    = true
+  bundle_url = "https://policy-server.example.com/bundles/tflint.tar.gz"
+}
+```
+
+To authenticate with the bundle server, set the `TFLINT_OPA_BUNDLE_TOKEN` environment variable. The value is sent as a Bearer token in the `Authorization` header. See [Environment Variables](./environment_variables.md) for details.
+
+Both `bundle_url` and `policy_dir` can be used together. When both are set, policies from the local directory take precedence over bundle policies if they share the same package path. This allows organizations to distribute shared policies via a bundle while teams add local overrides.
```
Comment (wata727) [inline docs/configuration.md]: In my understanding, all policies are placed in the `tflint` package, so merging should be the same as a replacement, but are there any situations where merging works?

File: docs/configuration.md [hunk 4]
```diff
@@ -24,3 +25,29 @@ Change the directory from which policies are loaded. The priority is as follows:
...
+
+### Caching
+
+When the bundle server supports [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) headers, the plugin caches the downloaded bundle locally to avoid re-downloading unchanged bundles on subsequent runs. The cache is stored at `~/.tflint.d/cache/bundles`. Caching is automatic and requires no configuration.
```
Comment (wata727) [inline docs/configuration.md]: `cache/bundles` is too generic. Please use `cache/opa/bundles` instead.

File: docs/environment_variables.md [hunk 5]
```diff
@@ -4,6 +4,10 @@ Below is a list of environment variables that have meaning in the OPA ruleset:
 
 - `TFLINT_OPA_POLICY_DIR`
   - Directory where policy files are placed. See [Configuration](./configuration.md).
+- `TFLINT_OPA_BUNDLE_URL`
+  - URL of a remote OPA bundle server. See [Configuration](./configuration.md).
+- `TFLINT_OPA_BUNDLE_TOKEN`
+  - Bearer token for authenticating with a remote bundle server. When set, the token is sent in the `Authorization: Bearer <token>` header when fetching a bundle via `bundle_url`. See [Configuration](./configuration.md).
```
Comment (wata727) [inline docs/environment_variables.md]: By the way, what technology stack is the team maintaining the OPA server expected to adopt? Intuitively, it seems like it would be not worthwhile to develop and host an HTTP server just for this purpose.

If a managed solution like S3 is available, it makes sense, but in that case, whether you need a bearer token or basic auth will depend on the technology stack you're considering.

File: integration/remote_bundle/policies/main.rego
Comment (wata727) [inline integration/remote_bundle/policies/main.rego:1]: Testing remote bundles seems a bit too complex. All we need to verify here is that the bundled policy fires; unknown or sensitive values ​​are not of interest.

Similarly, overrides by local policies could be covered in a single test case.

File: opa/bundle.go (hunk 1 above)
Comment (mostealth) [reply opa/bundle.go]: Good catch. Will fix by initializing remaining to `maxBundleSize + 1`, so bundles exactly at 256 MiB are accepted and the error only triggers when strictly exceeding the limit.

File: docs/configuration.md (hunk 4 above)
Comment (mostealth) [reply docs/configuration.md]: Agreed. Will change to `cache/opa/bundles`.

File: docs/environment_variables.md (hunk 5 above)
Comment (mostealth) [reply docs/environment_variables.md]: Bearer token was chosen as a common auth mechanism for HTTP-served OPA bundles. For managed solutions like S3 or GCS, pre-signed URLs work without any auth headers — `bundle_url` would just point to the pre-signed URL directly. That said, we are not going to use authentication, happy to remove auth support from this first version if you'd prefer to keep the initial scope smaller, and add it in a follow-up once the right auth schemes are clearer.

File: integration/remote_bundle/policies/main.rego
Comment (mostealth) [reply integration/remote_bundle/policies/main.rego:1]: Agreed. Will simplify to a single rule checking `instance_type.value != "t2.micro"` with a single-resource `main.tf`. Same simplification for the override test. This also eliminates the version-specific result file issue that caused the CI failures.

File: docs/configuration.md (hunk 3 above)
Comment (mostealth) [reply docs/configuration.md]: Correct — since all policies must use `package tflint` for the custom builtins to work, the package-path merge is effectively a full replacement in practice. Will update the docs to clarify this. The merge logic still handles the general case correctly, but the docs will explicitly state that local policies replace bundle policies when both are present. Data documents are a different story — those are deep-merged so teams can override specific values without losing bundle-provided defaults.

File: opa/ruleset.go (hunk 2 above)
Comment (mostealth) [reply opa/ruleset.go]: You're right. Will replace the store overwrite with a deep-merge (mergeData) — bundle data is preserved, local data values win on conflicts. This supports the use case where teams use bundle policies as-is but override specific data values locally (e.g., customizing data.config.allowed_types without touching the policies).

**Comment by mostealth** [comment]
Updated — removed `TFLINT_OPA_BUNDLE_TOKEN` and all auth support from this PR to keep the initial scope smaller. Auth can be added in a follow-up once the right scheme is agreed upon. Also addressed all other review comments (simplified tests, fixed limitedReader off-by-one, renamed cache dir to cache/opa/bundles, deep-merge data stores).

**Comment by mostealth** [comment]
@wata727 can we progress this PR?

**Comment by mostealth** [comment]
All review feedback is addressed and the branch is rebased on the latest main (single commit). I verified the full CI matrix passes by running the same workflows in my fork — including a Windows-specific path fix in the integration test: [fork checks](https://github.com/mostealth/tflint-ruleset-opa/pull/1/checks). I've also been running this against a real bundle server internally and it works end-to-end. This just needs the workflows approved to run CI here — could you click "Approve and run workflows" when you have a moment, @wata727?

File: docs/configuration.md [hunk 6]
```diff
@@ -24,3 +25,27 @@ Change the directory from which policies are loaded. The priority is as follows:
...
+2. `TFLINT_OPA_BUNDLE_URL` environment variable
+
+```hcl
+plugin "opa" {
+  enabled    = true
+  bundle_url = "https://policy-server.example.com/bundles/tflint.tar.gz"
+}
+```
+
+Both `bundle_url` and `policy_dir` can be used together. When both are set, local policies replace bundle policies that share the same package path. Since all policies use the `tflint` package, local policies effectively replace the entire bundle when both are present. Data documents (JSON/YAML) from both sources are deep-merged, with local values taking precedence on conflicts.
```
Comment (wata727) [inline docs/configuration.md]: This explanation feels a bit redundant. At least in this ruleset, all policies are replaced, so there's no need to mention the general statement "local policies replace bundle policies that share the same package path".

Also, is replacement truly the expected behavior? Ideally, shouldn't it be merged instead?
What are the typical behaviors of OPA tools that support other remote bundles when they conflict with local policies?

File: docs/configuration.md [hunk 7]
```diff
@@ -24,3 +25,27 @@ Change the directory from which policies are loaded. The priority is as follows:
...
+
+### Caching
+
+When the bundle server supports [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) headers, the plugin caches the downloaded bundle locally to avoid re-downloading unchanged bundles on subsequent runs. The cache is stored at `~/.tflint.d/cache/opa/bundles`. Caching is automatic and requires no configuration.
```
Comment (wata727) [inline docs/configuration.md]: Since the cache behavior is an implementation detail, I don't think it needs to be mentioned in the configuration documentation.

File: docs/configuration.md (hunk 6 above)
Comment (mostealth) [reply docs/configuration.md]: Good question — let me lay out the whole picture for both policies and data, since they're entangled.

**The problem.** When `bundle_url` and `policy_dir` are both set, we have to decide how bundle and local content combine on two axes: **rego modules** and **data documents**. The complication is that this ruleset requires every rule to be in `package tflint` (the custom builtins resolve `data.tflint.<rule>`), so bundle and local *always* overlap at the `data.tflint` root — they can never be disjoint.

**What other OPA tooling does (and why it doesn't transfer).** OPA's strategy is conflict avoidance via non-overlapping bundle roots: the bundle manager treats overlapping roots as an activation error; within a compiled set, partial rules union and  complete-rule / data-path clashes are errors. Conftest's `pull` just downloads policies into a directory and unions everything. Styra/Enterprise OPA compose `system` + `library` bundles, again via disjoint roots. None of them implement "local overrides remote" — they don't need to, because roots don't overlap. That escape hatch is exactly what our single-`package tflint` constraint removes, so there's no off-the-shelf precedent to borrow; we have to pick a semantic deliberately.

**Where the current code stands.** It's inconsistent across the two axes:
  - *Policies*: package-level replace — and since everything is `package tflint`, any local policy discards the **entire** bundle. That's the coarse behavior you flagged.
  - *Data*: deep-merged, with **local winning on conflicting paths** (this departs from OPA, which would error on overlapping data roots — but it's a reasonable convenience).

**Proposal.** Make policies behave like data already does: **merge the rule sets; on a same-name conflict, the local rule wins** and the bundle's copy is dropped, while all other bundle rules survive. Then both axes follow one rule — *"bundle and local are merged; on a conflict (same rule name, or same data path), local takes precedence"* — which is also a single doc sentence. Users who want to silence a bundle rule without redefining it still have TFLint's `rule "opa_deny_x" { enabled = false }`.

(Pure union — both versions fire — is the OPA-literal behavior, but it double-reports on collisions and would be inconsistent with how we already treat data, so I'm proposing local-precedence instead.)

If that sounds right to you, I'll switch policies to local-wins merge, update the override integration test (local redefines one rule; a second bundle-only rule still fires), and trim the redundant wording.

File: docs/configuration.md (hunk 7 above)
Comment (mostealth) [reply docs/configuration.md]: Agreed — I'll remove the Caching section from the configuration docs. The ETag caching itself stays; it's automatic and doesn't need to be surfaced as user-facing config.

File: docs/configuration.md (hunk 6 above)
Comment (wata727) [reply docs/configuration.md]: I recommend starting with the simplest approach. If other solutions don't allow package conflicts, shouldn't it throw an error if both remote and local policies are configured?

Comment (mostealth) [reply docs/configuration.md]: As you suggested, I went with the simplest approach: bundle_url and a local policy directory are now mutually exclusive. If a bundle is configured and policy_dir (or its default locations) also contains policies, the plugin errors out instead of merging or shadowing. An empty or data-only local directory isn't a conflict and is ignored in favor of the bundle.

This removes the merge/override logic entirely (including mergeData). I also trimmed the redundant docs wording and dropped the Caching section (implementation detail), removed the now-obsolete override integration test, and added unit tests for the conflict path.

Verified locally and via the full CI matrix on my fork (build + e2e on ubuntu/windows × tflint v0.46.0/latest, all green).

File: integration/remote_bundle/result.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
+{
+  "issues": [
+    {
+      "rule": {
+        "name": "opa_deny_not_t2_micro",
+        "severity": "error",
+        "link": "/policies/main.rego:5"
```
Comment (wata727) [inline integration/remote_bundle/result.json]: It would be a good idea to set the location so that it is clear that this policy was retrieved remotely.

Comment (mostealth) [reply integration/remote_bundle/result.json]: Good idea — done. Bundle-sourced rules now report their location jar-style, prefixing the in-bundle path with the bundle URL:

`https://policy-server.example.com/bundles/tflint.tar.gz!/policies/main.rego:5`

The `!/` separator makes it clear the policy lives inside a remote bundle rather than on the local filesystem. Local policies are unchanged (`policies/main.rego:5`), so the location still reflects where each rule actually came from. Updated the integration test (normalizing the test server's port) and the expected `result.json` accordingly.

File: README.md [hunk 9]
```diff
@@ -24,7 +24,19 @@ plugin "opa" {
...
+plugin "opa" {
+  enabled    = true
+  version    = "0.11.0"
+  source     = "github.com/terraform-linters/tflint-ruleset-opa"
+
+  bundle_url = "https://policy-server.example.com/bundles/tflint.tar.gz"
+}
+```
+
+For local policies, files are placed under `~/.tflint.d/policies` or `./.tflint.d/policies`. First create a directory:
```
Comment (wata727) [inline README.md]: In my intuition, since bundle policies are an advanced feature, there is no need to mention them in the README.

File: opa/bundle.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	if int64(len(p)) > l.remaining {
+		p = p[:l.remaining]
```
Comment (wata727) [inline opa/bundle.go]: Is there any reason to perform a partial read when the size exceeds `remaining`? Even if the read succeeds, wouldn't that result in reading a corrupted bundle?

File: README.md (hunk 9 above)
Comment (mostealth) [reply README.md]: Good call — removed the `bundle_url` section from the README; it's covered in `docs/configuration.md`.

Comment (mostealth) [reply README.md]: Good call — removed the `bundle_url` section from the README; it's covered in `docs/configuration.md`.

File: opa/bundle.go (hunk 10 above)
Comment (mostealth) [reply opa/bundle.go]: You're right. Replaced the custom `limitedReader` with `io.ReadAll(io.LimitReader(resp.Body, maxBundleSize+1))` + a length check — a truncated body is never used.
Keeps your earlier fix intact (exactly-at-limit accepted, only strictly larger rejected); added a boundary test to lock it in.

File: opa/ruleset.go [hunk 11]
```diff
@@ -40,23 +44,65 @@ func (r *RuleSet) ApplyConfig(body *hclext.BodyContent) error {
...
-	if err != nil {
-		return fmt.Errorf("failed to load policies; %w", err)
+	// When a bundle is configured, it becomes the policy source. Every rule must
+	// live in the `tflint` package for the custom builtins to resolve, so a bundle
+	// and local policies would always conflict; reject that combination instead of
+	// merging or shadowing. A local directory with no policies (empty, or data
+	// files only) is not a conflict and is simply ignored in favor of the bundle.
+	bundleURL := r.config.bundleURL()
+	if bundleURL != "" {
+		if len(modules) > 0 {
```
Comment (wata727) [inline opa/ruleset.go]: Since local data is ignored, it should be treated as an error here if such data exists.

File: opa/ruleset.go [hunk 12]
```diff
@@ -40,23 +44,65 @@ func (r *RuleSet) ApplyConfig(body *hclext.BodyContent) error {
...
+		}
+		modules = map[string]*ast.Module{}
+		for _, m := range b.Modules {
+			modules[m.Path] = m.Parsed
+		}
+		data = b.Data
+	}
+
+	if len(modules) == 0 {
+		return nil
```
Comment (wata727) [inline opa/ruleset.go]: Returning early when the module is empty leads to an issue where `ApplyGlobalConfig` cannot be called.

File: opa/ruleset.go [hunk 13]
```diff
@@ -40,23 +44,65 @@ func (r *RuleSet) ApplyConfig(body *hclext.BodyContent) error {
...
+		ret, err := loader.NewFileLoader().Filtered([]string{policyDir}, nil)
+		if err != nil {
+			return fmt.Errorf("failed to load policies; %w", err)
+		}
+		for k, m := range ret.ParsedModules() {
+			modules[k] = m
+		}
+		if len(ret.Documents) > 0 {
+			data = ret.Documents
 		}
```
Comment (wata727) [inline opa/ruleset.go]: Wouldn't it be sufficient to simply assign it?


```suggestion
		modules = ret.ParsedModules()
		data = ret.Documents
}
```

File: opa/ruleset.go (hunk 11 above)
Comment (mostealth) [reply opa/ruleset.go]: Good point — done. Any local content now conflicts: `bundle_url` + a local dir containing policies **or data** errors out (previously data-only was silently ignored). Updated the check, the docs, and added a test.

File: opa/ruleset.go (hunk 12 above)
Comment (mostealth) [reply opa/ruleset.go]: Fixed — rule generation is now guarded behind a "policies present" check and `ApplyGlobalConfig` is always called at the end, so it's no longer skipped when there are no modules.

File: opa/ruleset.go (hunk 13 above)
Comment (mostealth) [reply opa/ruleset.go]: Done — replaced the copy loop with the direct assignment as suggested.

**Review by wata727 — APPROVED** [review]
Thanks!
